sips/sip-015/sip-015-network-upgrade.md

```diff
@@ -0,0 +1,1582 @@
+
+**Rationale:** This method makes it possible to set values of a sequence type
+without incurring the costs of splitting the sequence into subsequences,
+and concatenating them back together with the new value.  This operation is also
+a common facility in many other languages, so supporting it in Clarity would
+make Clarity more approachable to new developers.
+
+**Examples:**
+
+```clarity
+(replace-at u"ab" u1 u"c") ;; Returns (some u"ac")
+(replace-at 0x00112233 u2 0x44) ;; Returns (some 0x00114433)
+(replace-at "abcd" u3 "e") ;; Returns (some "abce")
+(replace-at (list 1) u0 10) ;; Returns (some (10))
+(replace-at (list (list 1) (list 2)) u0 (list 33)) ;; Returns (some ((33) (2)))
+(replace-at (list 1 2) u3 4) ;; Returns none
+```
+
+### New global: `tx-sponsor?`
+
+* **Type:** `(optional principal)`
+
+This global variable evaluates to the fee-sponsoring principal of the current transaction (if there is such a principal).
+
+**Rationale:** Stacks already exposes the origin address as `tx-sender`, and
+exposing this data would make it possible to build smart contracts where only
+certain entities could relay transactions on behalf of users (such as entities
+with which the user has an off-chain business relationship).
+
+### New global: `chain-id`
+
+* **Type:** `uint`
+
+This global variable evaluates to a 32-bit chain ID that identifies this
+instance of the Stacks blockchain.  The purpose of this global variable is to
+give contract developers a way to determine which instance of Stacks their code
+runs on, in order to execute instance-specific logic or integrate with
+instance-specific off-chain services.
+
+The following values are reported:
+
+* `u1`:  This is the Stacks mainnet chain (this is `0x00000001`).
+
+* `u2147483648`:  This is the Stacks testnet chain (this is `0x80000000`).
+
+Other values may be used for other deployments of Stacks, such as (but not
+limited to) subnets.
+
+**Rationale**: Stacks already has two instances (testnet and mainnet), and
+subnets are instances of Stacks.  Enabling smart contracts to differentiate
+between which environments they run in would empower developers to add
+(or rely on) chain-specific features.
+
+### New feature: Faster Clarity Parser
+
+This SIP proposes a new Clarity lexer and parser implementation that will be
+specific to Clarity 2.  Current benchmarks indicate that the reference
+implementation of this new lexer and parser it is around 3x faster than the
+reference implementation of the chain today.
+
+The reason for proposing this as a breaking change is to de-risk the possibility
+that the new lexer and parser are not "bug-for-bug" compatible with the current
+parser.  Tying the release of this new lexer and parser with a breaking change,
+while keeping the old parser for processing all blocks prior to the change,
+ensures that any new behaviors introduced will only apply to transactions after
+the change takes effect.
+
+### New feature: New Trait Semantics
```

Comment:
Tagging @obycode to make sure this section accurately describes the Clarity 2 trait semantics.  You're the expert, not me!

Reply:
I see what you're going for with this description, but I don't think of it this way, and I find this explanation a bit confusing.

I think of the new behavior as almost the opposite: "**Traits are real types**". The proposed behavior treats traits more like a real type, in that they can be used in the same way as any other types, aside from persistence (traits may not be stored in a data-var or map), since this inhibits static analysis.

Below is how I would rewrite this section, which also adds some additional changes to the behavior.

---

Over the lifetime of Clarity 1, it was discovered that there are a few surprising behaviors of the current trait semantics.  Clarity 2 proposes the following trait semantics:

* **Traits are real types**  The Clarity2 type system will analyze and propagate the type of a trait value in the same ways that it does with other types. This means that in Clarity2, trait values can be embedded into complex types (lists, tuples, optionals, and responses) and maintain their ability to enable dynamic dispatch. The only limitation on the trait type is that trait values may not be persisted (e.g. in a data-var or map). Persisting a trait value, and maintaining the ability to perform dynamic dispatch would inhibit static analysis of a contract.

* **Trait values may be coerced into compatible trait types**  When type checking a trait value against another trait type, the expected trait type need not be the identical trait type of the value; instead, the trait value is checked for compatibility with the expected type, meaning that if the expected trait type is a subset of the trait value's type, then the coercion is legal. See below for an example.

* **Trait values may be locally bound**  A trait value can be bound in a local scope and maintain its trait type and dynamic dispatch functionality. Binding in a local scope includes `let` expressions and `match` expressions.

* **Traits may not be defined with duplicate function names**  In Clarity1 a trait could be defined with duplicated function names in its signature. This resulted in only the last instance (in program order) being kept as part of the trait's definition. In Clarity2, a trait defined with a duplicate function name will trigger a `DefineTraitDuplicateMethod` error.

* **Contract principals stored in constants are callable** Clarity2 allows a constant contract principal (defined with a `define-constant`) to be callable -- used to make static dispatch contract calls, or coerced into a trait type.

* **Imported trait name conflicts** In Clarity1, there is surprising behavior relating to the names of imported traits (from `use-trait`). The name of a local trait will conflict with the trait name of the imported trait. For example, in `(use-trait a-alias .a-trait.a)` if a local trait is defined with the name `a`, then uses of `a-alias` will refer to the local trait, `a` instead of the imported trait. In Clarity2, the imported trait can always be referenced by its alias and the imported trait name will not conflict with local traits.

Reply:
Thanks; added :+1: